### PR TITLE
IDENTITY-6654: Unit test coverage for OIDCSession management

### PR DESCRIPTION
**Pull request by piraveena**

### CODE REVIEW

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.mockito.Mock;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.wso2.carbon.identity.oidc.session.cache.OIDCSessionParticipantCache;
+import org.wso2.carbon.identity.oidc.session.servlet.TestOIDCSessionBase;
+import org.wso2.carbon.identity.oidc.session.servlet.TestUtil;
+
+import static org.testng.Assert.*;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java]: Don't use * imports

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @Mock
+    OIDCSessionState oidcSessionState;
+
+    private OIDCSessionManager oidcSessionManager;
+    private static final String SESSION_ID = "090907ce-eab0-40d2-a46d-acd4bb33f0d0";
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        oidcSessionManager=new OIDCSessionManager();
+        TestUtil.startTenantFlow("carbon.super");
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java]: Do you always need to have a tenant flow started before all tests ? If not only start the tenant flow whenever required.

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        when(request.getParameter(OIDCSessionConstants.OIDC_REDIRECT_URI_PARAM)).thenReturn(redirectUri);
+
+        mockStatic(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(oAuthServerConfiguration);
+        when(oAuthServerConfiguration.getPersistenceProcessor()).thenReturn(tokenPersistenceProcessor);
+        when(tokenPersistenceProcessor.getProcessedClientId(anyString())).thenAnswer(new Answer<Object>() {
+
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java]: Remove unnecessary new line

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        when(OIDCSessionManagementUtil.getOrigin((CALLBACK_URL))).thenReturn("http://localhost:8080/playground2");
+
+        mockStatic(IdentityDatabaseUtil.class);
+        when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+
+        PrintWriter out = new PrintWriter("oidc-session.txt");
+        when(response.getWriter()).thenReturn(out);
+        oidcSessionIFrameServlet.doGet(request, response);
+        out.flush();
+        assertTrue(FileUtils.readFileToString(new File("oidc-session.txt"), "UTF-8").contains(expected));
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java]: Do  you really need to write this to a file ? . Can't you assert without writing to a file ? . If this is required where are u creating the file ? location ?

File: components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml [hunk 5]
```diff
@@ -0,0 +1,684 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml]: Are you sure that a carbon.xml file is required ? I haven't come across any case where carbon.xml file is required

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java (hunk 2 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java]: Yes. I need

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java (hunk 4 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java]: I can't assert without writing to a file. The test method is a void method and the response is written using a print writer in the test method. That is (response.getWriter().print(ERROR_RESPONSE))

File: components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml (hunk 5 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml]: I'm getting this error when I don't have carbon.xml file. 
32 [main] ERROR org.wso2.carbon.utils.CarbonUtils - Cannot read file /home/piraveena/Documents/identity-inbound-oauth-unit-test/identity-inbound-auth-oauth/components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml
java.io.FileNotFoundException: /home/piraveena/Documents/identity-inbound-oauth-unit-test/identity-inbound-auth-oauth/components/org.wso2.carbon.identity.oidc.session/src/test/resources/repository/conf/carbon.xml (No such file or directory)

**Comment by piraveena** [comment]
Fixed with requested changes.

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java [hunk 6]
```diff
@@ -29,12 +29,20 @@
...
         System.setProperty("carbon.home", carbonHome);
         PrivilegedCarbonContext.startTenantFlow();
         PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain);
     }
 
+    public static void endTenantFlow(String tenantDomain) {
+        String carbonHome = TestUtil.class.getResource("/").getFile();
+        System.setProperty("carbon.home", carbonHome);
+        PrivilegedCarbonContext.endTenantFlow();
+       // PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java]: Why is this line commented?

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java [hunk 7]
```diff
@@ -29,12 +29,20 @@
...
+        String carbonHome = Paths.get(System.getProperty("user.dir"), "src", "test", "resources").toString();
+
         System.setProperty("carbon.home", carbonHome);
         PrivilegedCarbonContext.startTenantFlow();
         PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain);
     }
 
+    public static void endTenantFlow(String tenantDomain) {
+        String carbonHome = TestUtil.class.getResource("/").getFile();
+        System.setProperty("carbon.home", carbonHome);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java]: Better to use `CarbonBaseConstants.CARBON_HOME`

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java [hunk 8]
```diff
@@ -29,12 +29,20 @@
 public class TestUtil {
 
     public static void startTenantFlow(String tenantDomain) {
-        String carbonHome = TestUtil.class.getResource("/").getFile();
+        String carbonHome = Paths.get(System.getProperty("user.dir"), "src", "test", "resources").toString();
+
         System.setProperty("carbon.home", carbonHome);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/TestUtil.java]: Better to use `CarbonBaseConstants.CARBON_HOME`

File: components/org.wso2.carbon.identity.oidc.session/pom.xml [hunk 9]
```diff
@@ -115,6 +115,10 @@
             <artifactId>h2</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oidc.session/pom.xml]: Do we need this dependency?

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    OIDCSessionState oidcSessionState;
+
+    private OIDCSessionManager oidcSessionManager;
+    private static final String SESSION_ID = "090907ce-eab0-40d2-a46d-acd4bb33f0d0";
+    private static final String NEW_SESSION_ID = "080907ce-eab0-40d2-a46d-acd4bb33f0d0";
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        oidcSessionManager=new OIDCSessionManager();
+        TestUtil.startTenantFlow("carbon.super");
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java]: Better to use `MultitenantConstants.SUPER_TENANT_DOMAIN_NAME`

**Comment by piraveena** [comment]
Added requested changes

File: components/org.wso2.carbon.identity.oidc.session/pom.xml (hunk 9 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oidc.session/pom.xml]: For PowerMockIdentityBaseTest , we need this dependency

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    @Mock
+    OIDCSessionState oidcSessionState;
+
+    private OIDCSessionManager oidcSessionManager;
+    private static final String SESSION_ID = "090907ce-eab0-40d2-a46d-acd4bb33f0d0";
+    private static final String NEW_SESSION_ID = "080907ce-eab0-40d2-a46d-acd4bb33f0d0";
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        oidcSessionManager=new OIDCSessionManager();
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManagerTest.java]: Add one line after method signature

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServletTest.java [hunk 12]
```diff
@@ -94,12 +97,20 @@
...
+    OAuthAppDO oAuthAppDO;
+
+    @Mock
+    KeyStore keyStore;
+
     private static final String CLIENT_ID_VALUE = "3T9l2uUf8AzNOfmGS9lPEIsdrR8a";
     private static final String APP_NAME = "myApp";
     private static final String SECRET = "87n9a540f544777860e44e75f605d435";
     private static final String USERNAME = "user1";
+    private static final String CALLBACK_URL="http://localhost:8080/playground2/oauth2client";
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServletTest.java]: Add spaces between =

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServletTest.java [hunk 13]
```diff
@@ -133,47 +144,69 @@ public void setUp() throws Exception {
...
+                {opbsCookie, true, redirectUrl[7], "Post", " ", null, false,
+                        idTokenHint, false, INVALID_CALLBACK_URL},
         };
     }
 
     @Test(dataProvider = "provideDataForTestDoGet")
     public void testDoGet(Object cookie, boolean sessionExists, String redirectUrl, String expected, String consent,
                           String sessionDataKey, boolean skipUserConsent, String idTokenHint,
-                          boolean isJWTSignedWithSPKey) throws Exception {
+                          boolean isJWTSignedWithSPKey, String postLogoutUrl ) throws Exception {
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServletTest.java]: Add line after signature

File: components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    private static final String SECRET__WITH_NO_CALLBACK_URL = "87n9a540f544777860e44e75f605d445";
+    private static final String CLIENT_ID_WITH_VALID_REGEX_CALLBACK_URL = "4T9l2uUf8AzNOfmGS9lPEIsdrR7b";
+    private static final String CLIENT_ID_WITH_INVALID_REGEX_CALLBACK_URL = "4T8l2uUf8AzNOfmGS9lPEIsdrR7b";
+    private static final String INVALID_REGEX_CALLBACK_URL = "regexp=http://localhost:8080/playground" +
+            ".appone/oauth2client";
+    private static final String VALID_REGEX_CALLBACK_URL = "regexp=http://localhost:8080/playground2/oauth2client";
+
+    @BeforeTest
+    public void setUp() throws Exception {
+        oidcSessionIFrameServlet = new OIDCSessionIFrameServlet();
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.oidc.session/src/test/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServletTest.java]: Add line after signature

**Pull request merged**